<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>process Eigenstrat data - alternative version — processEigenstrat • BREADR</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="process Eigenstrat data - alternative version — processEigenstrat"><meta name="description" content="A function that takes paths to an eigenstrat trio (ind, snp and geno file) and
returns the pairwise mismatch rate for all pairs on a thinned set of SNPs.
Options include choosing thinning parameter, subsetting by population names,
and filtering out SNPs for which deamination is possible."><meta property="og:description" content="A function that takes paths to an eigenstrat trio (ind, snp and geno file) and
returns the pairwise mismatch rate for all pairs on a thinned set of SNPs.
Options include choosing thinning parameter, subsetting by population names,
and filtering out SNPs for which deamination is possible."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">BREADR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.3.9000</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/example-pipeline.html">Example pipeline</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/jonotuke/BREADR/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>process Eigenstrat data - alternative version</h1>
      <small class="dont-index">Source: <a href="https://github.com/jonotuke/BREADR/blob/HEAD/R/processEigenstrat.R" class="external-link"><code>R/processEigenstrat.R</code></a></small>
      <div class="d-none name"><code>processEigenstrat.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>A function that takes paths to an eigenstrat trio (ind, snp and geno file) and
returns the pairwise mismatch rate for all pairs on a thinned set of SNPs.
Options include choosing thinning parameter, subsetting by population names,
and filtering out SNPs for which deamination is possible.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">processEigenstrat</span><span class="op">(</span></span>
<span>  <span class="va">indfile</span>,</span>
<span>  <span class="va">genofile</span>,</span>
<span>  <span class="va">snpfile</span>,</span>
<span>  filter_length <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  pop_pattern <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  filter_deam <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  outfile <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  chromosomes <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-indfile">indfile<a class="anchor" aria-label="anchor" href="#arg-indfile"></a></dt>
<dd><p>path to eigenstrat ind file</p></dd>


<dt id="arg-genofile">genofile<a class="anchor" aria-label="anchor" href="#arg-genofile"></a></dt>
<dd><p>path to eigenstrat geno file.</p></dd>


<dt id="arg-snpfile">snpfile<a class="anchor" aria-label="anchor" href="#arg-snpfile"></a></dt>
<dd><p>path to eigenstrat snp file.</p></dd>


<dt id="arg-filter-length">filter_length<a class="anchor" aria-label="anchor" href="#arg-filter-length"></a></dt>
<dd><p>the minimum distance between sites to be compared (to reduce the effect of LD).</p></dd>


<dt id="arg-pop-pattern">pop_pattern<a class="anchor" aria-label="anchor" href="#arg-pop-pattern"></a></dt>
<dd><p>a character vector of population names to filter the ind file if only some populations are to compared.</p></dd>


<dt id="arg-filter-deam">filter_deam<a class="anchor" aria-label="anchor" href="#arg-filter-deam"></a></dt>
<dd><p>a TRUE/FALSE for if C-&gt;T and G-&gt;A sites should be ignored.</p></dd>


<dt id="arg-outfile">outfile<a class="anchor" aria-label="anchor" href="#arg-outfile"></a></dt>
<dd><p>(OPTIONAL) a path and filename to which we can save the output of the function as a TSV,
if NULL, no back up saved. If no outfile, then a tibble is returned.</p></dd>


<dt id="arg-chromosomes">chromosomes<a class="anchor" aria-label="anchor" href="#arg-chromosomes"></a></dt>
<dd><p>the chromosome to filter the data on.</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>controls printing of messages to console</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>out_tibble: A tibble containing four columns:</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># Use internal files to the package as an example</span></span></span>
<span class="r-in"><span><span class="va">indfile</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"example.ind.txt"</span>, package <span class="op">=</span> <span class="st">"BREADR"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">genofile</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"example.geno.txt"</span>, package <span class="op">=</span> <span class="st">"BREADR"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">snpfile</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"example.snp.txt"</span>, package <span class="op">=</span> <span class="st">"BREADR"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">processEigenstrat</span><span class="op">(</span></span></span>
<span class="r-in"><span><span class="va">indfile</span>, <span class="va">genofile</span>, <span class="va">snpfile</span>,</span></span>
<span class="r-in"><span>filter_length<span class="op">=</span><span class="fl">1e5</span>,</span></span>
<span class="r-in"><span>pop_pattern<span class="op">=</span><span class="cn">NULL</span>,</span></span>
<span class="r-in"><span>filter_deam<span class="op">=</span><span class="cn">FALSE</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Reading in SNP data.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Analysing chromosomes:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Starting to read in genotype data.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   |                                                                              |                                                                      |   0%  |                                                                              |==============                                                        |  20%  |                                                                              |============================                                          |  40%  |                                                                              |==========================================                            |  60%  |                                                                              |========================================================              |  80%  |                                                                              |======================================================================| 100%  Complete.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Starting to compare genotypes and calculate PMR.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   |                                                                              |                                                                      |   0%  |                                                                              |=====                                                                 |   7%  |                                                                              |==========                                                            |  14%  |                                                                              |===============                                                       |  21%  |                                                                              |====================                                                  |  29%  |                                                                              |=========================                                             |  36%  |                                                                              |==============================                                        |  43%  |                                                                              |===================================                                   |  50%  |                                                                              |========================================                              |  57%  |                                                                              |=============================================                         |  64%  |                                                                              |==================================================                    |  71%  |                                                                              |=======================================================               |  79%  |                                                                              |============================================================          |  86%  |                                                                              |=================================================================     |  93%  |                                                                              |======================================================================| 100%</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Complete.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># A tibble: 15 × 4</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    pair        nsnps mismatch    pmr</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 1</span> Ind1 - Ind2     2        1  0.5  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 2</span> Ind1 - Ind3    14        2  0.143</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 3</span> Ind1 - Ind4    14        3  0.214</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 4</span> Ind1 - Ind5     1        0  0    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 5</span> Ind1 - Ind6     4        0  0    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 6</span> Ind2 - Ind3    13        3  0.231</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 7</span> Ind2 - Ind4    11        2  0.182</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 8</span> Ind2 - Ind5     0        0 <span style="color: #BB0000;">NA</span>    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 9</span> Ind2 - Ind6     4        2  0.5  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">10</span> Ind3 - Ind4    35        5  0.143</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">11</span> Ind3 - Ind5    10        0  0    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">12</span> Ind3 - Ind6    19        2  0.105</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">13</span> Ind4 - Ind5    10        1  0.1  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">14</span> Ind4 - Ind6    20        7  0.35 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">15</span> Ind5 - Ind6     2        1  0.5  </span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Jono Tuke, Adam B. Rohrlach, Wolfgang Haak, Divyaratan Popli.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

